### ISSUE TITLE
API Gateway CORS is blocking localhost access

### ISSUE THREAD

**Issue by andrecasal** [aws-amplify/amplify-cli#8019]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm i -g @aws-amplify/cli

### If applicable, what version of Node.js are you using?

16.6.0

### Amplify CLI Version

5.3.0

### What operating system are you using?

macOS 11.5.2

### Amplify Categories

api

### Amplify Commands

add, push, update

### Describe the bug

After running amplify update API and adding access restrictions to the API endpoint for only authenticated users, I can no longer access the lambda function from my localhost.

After running this:
<img width="604" alt="Screenshot 2021-08-24 at 11 35 10" src="https://user-images.githubusercontent.com/22457463/130602502-83848f4d-6c09-446a-9bc8-2e24ccd10c35.png">

This happens:
<img width="1673" alt="Screenshot 2021-08-24 at 11 35 38" src="https://user-images.githubusercontent.com/22457463/130602539-19a9fcac-f6c8-4d7c-a55f-b1511077c31e.png">

### Expected behavior

I would expect that place to edit the CORS policies to be obvious or easier to find. I've searched the AWS API Gateway console but I found nothing related to CORS policies anywhere.

### Reproduction steps

1. Run `amplify ad api` and add some lambda function
2. Refer to the "Describe the bug section" and run that command

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by andrecasal** [comment]
Could this be a CloudFront problem?

<img width="1278" alt="Screenshot 2021-08-24 at 12 03 20" src="https://user-images.githubusercontent.com/22457463/130605913-cf231f54-94dd-4485-8191-c3d5fa99fc62.png">

**Comment by RoniqueRicketts** [comment]
I have the same issue I am on the same version on Amplify (the same issue was on the previous version also). I am on Windows 10 OS. What I found out for this issue so far:
1. The request never reaches the Lambda functions connected to the API.
2. If I turn off Authenticated users only I am able to hit the open API with the same failed frontend request from before and get a 200 response.
3. Disabling the IAM Feature off the endpoint makes the API accessible which gives me 200 response but the API is not protected.
4. If I update the policy that controls the authRole to full access I get the same error.

**Comment by andrecasal** [comment]
This is weird because the preflight request says I can POST from any URL:
<img width="778" alt="Screenshot 2021-08-25 at 12 00 47" src="https://user-images.githubusercontent.com/22457463/130779732-512ab582-aad4-4d1a-9437-309df2fae470.png">

But the request itself fails with AccessDeniedException:
<img width="621" alt="Screenshot 2021-08-25 at 12 02 51" src="https://user-images.githubusercontent.com/22457463/130779803-a82797b5-c683-4d12-b80a-4ef4cd0ec1ed.png">

**Comment by josefaidt** [comment]
Hey y'all :wave: after restricting access to the API route, I was able to add the CORS headers (or uncomment what's provided from the templates) to mitigate this issue 

```diff
exports.handler = async (event) => {
  // TODO implement
  const response = {
    statusCode: 200,
    //  Uncomment below to enable CORS requests
+   headers: {
+     'Access-Control-Allow-Origin': '*',
+     'Access-Control-Allow-Headers': '*',
+   },
    body: JSON.stringify('Hello from Lambda!'),
  }
  return response
}
```

and my sample app.js file in a Next.js app making the call with authentication

```js
// src/pages/_app.js
import { useEffect, useState } from 'react'
import { Amplify, API } from 'aws-amplify'
import { withAuthenticator, AmplifySignOut } from '@aws-amplify/ui-react'
import config from '../aws-exports.js'

Amplify.configure(config)

async function get() {
  const apiName = 'api8019'
  const path = '/hello'
  const myInit = {
    response: true,
  }

  return API.get(apiName, path, myInit)
    .then((response) => {
      console.log('got response', response)
      return response
    })
    .catch((error) => {
      console.log('got error', error)
      return error
    })
}

function App() {
  const [response, setResponse] = useState()
  const [isLoading, setIsLoading] = useState(false)
  async function getResponse() {
    setIsLoading(true)
    setResponse(await get())
    setIsLoading(false)
  }
  useEffect(() => {
    getResponse()
  }, [])
  return (
    <div>
      <AmplifySignOut />
      <button onClick={getResponse}>refresh</button>
      {isLoading && <span>Loading...</span>}
      <code>
        <pre>{JSON.stringify(response, null, 2)}</pre>
      </code>
    </div>
  )
}

export default withAuthenticator(App)
```

**Comment by andrecasal** [comment]
@josefaidt I've tried that but my request is being blocked before it hits the function. I know that because when I remove authorization, I can see logs for the Lambda function and when I add authorization and get the error, those logs don't show up.

How did you add the API and function? Did you run `amplify add api`?

**Comment by josefaidt** [comment]
Hey @andrecasal yes I used `amplify add api` and added a GET and POST route. 

1. add the API with an unsecured route to confirm no CORS issues
   ```
    > amplify add api
    ? Please select from one of the below mentioned services: REST
    ? Provide a friendly name for your resource to be used as a label for this category in the project: api8019
    ? Provide a path (e.g., /book/{isbn}): /hello
    ? Choose a Lambda source Create a new Lambda function
    ? Provide an AWS Lambda function name: 8019hello
    ? Choose the runtime that you want to use: NodeJS
    ? Choose the function template that you want to use: Hello World
   ```
2. secure the route
   ```
    > amplify update api
    ? Please select from one of the below mentioned services: REST
    ? Please select the REST API you would want to update api8019
    ? What would you like to do Update path
    ? Please select the path you would want to edit /hello
    ? Provide a path (e.g., /book/{isbn}): /hello
    ? Choose a Lambda source Use a Lambda function already added in the current Amplify project
    ? Choose the Lambda function to invoke by this path 8019hello
    ? Restrict API access Yes
    ? Who should have access? Authenticated users only
    ? What kind of access do you want for Authenticated users? read
   ```
3. modify the Lambda handler to include CORS headers
   ```js
    exports.handler = async (event) => {
      // TODO implement
      const response = {
        statusCode: 200,
        //  Uncomment below to enable CORS requests
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Headers': '*',
        },
        body: JSON.stringify('Hello from Lambda!'),
      }
      return response
    }
   ```
4. add a POST route
   ```
    > amplify update api
    ? Please select from one of the below mentioned services: REST
    ? Please select the REST API you would want to update api8019
    ? What would you like to do Add another path
    ? Provide a path (e.g., /book/{isbn}): /hello-post
    ? Choose a Lambda source Create a new Lambda function
    ? Provide an AWS Lambda function name: 8019hellopost
    ? Choose the runtime that you want to use: NodeJS
    ? Choose the function template that you want to use: Hello World
    ...
    ? Restrict API access Yes
    ? Who should have access? Authenticated users only
    ? What kind of access do you want for Authenticated users? create
   ```
5. modify the Lambda handler to respond to POST requests and include CORS headers
   ```js
    exports.handler = async (event) => {
      // TODO implement
      const { name } = JSON.parse(event.body)
      const body = JSON.stringify(`Hello, ${name}!`)
      const response = {
        statusCode: 200,
        //  Uncomment below to enable CORS requests
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Headers': '*',
        },
        body,
      }
      return response
    }
   ```
6. run `amplify push`
7. in my frontend Next.js _app.js file:
   <details>
    <summary>code</summary>

    ```js
    import { useEffect, useState } from 'react'
    import { Amplify, API, Auth } from 'aws-amplify'
    import { withAuthenticator, AmplifySignOut } from '@aws-amplify/ui-react'
    import config from '../aws-exports.js'

    Amplify.configure(config)

    async function get() {
      const apiName = 'api8019'
      const path = '/hello'
      const myInit = {
        // response: true,
      }

      return API.get(apiName, path, myInit)
        .then((response) => {
          console.log('got response', response)
          return response
        })
        .catch((error) => {
          console.log('got error', error)
          return error
        })
    }

    async function post() {
      const apiName = 'api8019'
      const path = '/hello-post'
      const myInit = {
        // response: true,
        body: { name: 'World' },
      }

      return API.post(apiName, path, myInit)
        .then((response) => {
          console.log('got response', response)
          return response
        })
        .catch((error) => {
          console.log('got error', error)
          return error
        })
    }

    function App() {
      const [response, setResponse] = useState()
      const [isLoading, setIsLoading] = useState(false)
      async function getResponse() {
        setIsLoading(true)
        const getResponse = await get()
        setResponse((existing) => ({ ...existing, get: getResponse }))
        setIsLoading(false)
      }
      async function postResponse() {
        setIsLoading(true)
        const postResponse = await post()
        setResponse((existing) => ({ ...existing, post: postResponse }))
        setIsLoading(false)
      }
      function refresh() {
        getResponse()
        postResponse()
      }
      useEffect(() => {
        refresh()
      }, [])
      return (
        <div>
          <AmplifySignOut />
          <button onClick={refresh}>refresh</button>
          {isLoading && <span>Loading...</span>}
          <code>
            <pre>{JSON.stringify(response, null, 2)}</pre>
          </code>
        </div>
      )
    }

    export default withAuthenticator(App)
    ```

   </details>

8. run the Next.js development server with `yarn next` and authenticate
   
   ![image](https://user-images.githubusercontent.com/5033303/130801159-0fd041e4-a399-4598-b6fe-88a5b0748f06.png)

What is the status code you're receiving on the failed POST call?

**Comment by RoniqueRicketts** [comment]
@josefaidt I tried your above suggestion line by line and it didn't work I am still getting a cors error. I've even names the files the same way you've named yours. I am getting the error Cors Error - ```Cross-Origin Resource Sharing error: MissingAllowOriginHeader```. This said that x-cache ``Error from cloudfront``` and ```x-amzn-errortype: AccessDeniedException```. When I check cloudfront I see a policy that says 
```
Headers - Include the following headers
origin
access-control-request-headers
access-control-request-method
```
Should that origin be ```Access-Control-Allow-Origin```

**Comment by josefaidt** [comment]
@andrecasal what is the `aws-manual-config` import? We should be able to import the exports directly and configure. Can you paste the contents of your `aws-exports.js` file, omitting sensitive information?

```js
// src/aws-exports.js
const awsmobile = {
    "aws_project_region": "us-east-1",
    "aws_cognito_identity_pool_id": "us-east-1:7ac69283-cd4a-4b30-a5ae-d333d35204d5",
    "aws_cognito_region": "us-east-1",
    "aws_user_pools_id": "us-east-1_uHLbyAykk",
    "aws_user_pools_web_client_id": "703tihquc418thtstp62d4l9t9",
    "oauth": {},
    "aws_cloud_logic_custom": [
        {
            "name": "api8019two",
            "endpoint": "https://e2du5do64c.execute-api.us-east-1.amazonaws.com/dev",
            "region": "us-east-1"
        }
    ]
};


export default awsmobile;
```

@RoniqueRicketts how are you calling the API? What is the status code you're receiving on the failed call?

**Comment by andrecasal** [comment]
I just clone the project and ran `amplify pull --appId <appid> --envName production`. The aws-exports.js:
```
/* eslint-disable */
// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.

const awsmobile = {
    "aws_project_region": "us-east-2",
    "aws_cognito_identity_pool_id": "us-east-2:086e60d4-4f55-4444-94bc-41c22",
    "aws_cognito_region": "us-east-2",
    "aws_user_pools_id": "us-east-2_uOLq7",
    "aws_user_pools_web_client_id": "46gbgpvr9nlnc8aie4h",
    "oauth": {},
    "aws_cloud_logic_custom": [
        {
            "name": "stripeAPI",
            "endpoint": "https://8tm5.execute-api.us-east-2.amazonaws.com/production",
            "region": "us-east-2"
        }
    ]
};

export default awsmobile;
```

My `aws-manual-config.js` file:
```
const manualconfig = {
	endpoints: [
		{
			name: 'stripeAPI',
			endpoint: 'https://8tm5.execute-api.us-east-2.amazonaws.com/production'
		},
	]
}

export default manualconfig
```

**Comment by josefaidt** [comment]
Thanks @andrecasal for the information, are you able to import just the generated `aws-exports.js` file and configure Amplify?

```js
import { Amplify } from 'aws-amplify'
import config from '../aws-exports.js'

Amplify.configure(config)
```

**Comment by RoniqueRicketts** [comment]
@josefaidt That's the thing there is not status code. it just says Cors error
![cors error](https://user-images.githubusercontent.com/17409035/130817333-f5b86d7c-da13-4880-8b45-b6ef13ac68bb.PNG)
This is what it looks like.

**Comment by andrecasal** [comment]
> Thanks @andrecasal for the information, are you able to import just the generated `aws-exports.js` file and configure Amplify?
> 
> ```js
> import { Amplify } from 'aws-amplify'
> import config from '../aws-exports.js'
> 
> Amplify.configure(config)
> ```

Yeah, I cloned my project, deleted the `amplify` folder and the `aws-exports.js` file, ran `amplify pull --appId <appid> --envName production` and change the code to:
```
import config from 'aws-exports'
Amplify.configure(config)
```

**Comment by andrecasal** [comment]
I think @RoniqueRicketts and me have the exact same problem.

**Comment by andrecasal** [comment]
> @RoniqueRicketts and @andrecasal do y'all have Firefox installed and can look at the request there? Their dev tools will show the status

I'm on it.

**Comment by RoniqueRicketts** [comment]
Error 403

**Comment by andrecasal** [comment]
Oh, it shows up here (403 Forbidden):
<img width="1672" alt="Screenshot 2021-08-25 at 16 25 57" src="https://user-images.githubusercontent.com/22457463/130819237-3a6c5eda-d1f8-46cf-99ad-f3ae31f33b94.png">

**Comment by josefaidt** [comment]
Ah thank you @andrecasal and @RoniqueRicketts ! This error appears to be common in a failed call, for instance if we make a POST call to a GET only route. 

What dependencies are installed for the frontend? I'm installing the following:

```
yarn add next react react-dom classnames @aws-amplify/ui-react aws-amplify
```

**Comment by andrecasal** [comment]
It's private but I can give you access.

**Comment by andrecasal** [comment]
I can't find you on GitHub 🤔
<img width="397" alt="Screenshot 2021-08-25 at 17 21 47" src="https://user-images.githubusercontent.com/22457463/130828027-9437d93d-3e8d-4fb2-b250-d111541ff6ee.png">

**Comment by josefaidt** [comment]
@andrecasal try searching without the leading `@`

**Comment by josefaidt** [comment]
I've pushed my Next.js example to its own repo https://github.com/josefaidt/8019

**Comment by andrecasal** [comment]
> @andrecasal try searching without the leading `@`

I've invited you now 👍

**Comment by andrecasal** [comment]
> @andrecasal try searching without the leading `@`

I've invited you now 👍

**Comment by andrecasal** [comment]
I'm trying to rebuild the whole backend from scratch on a clone project. I forgot to `npm install stripe` on the Lambda function for this clone. It's funny that the exact same CORS error shows up. The only difference is the `x-amzn-errortype` header.

<img width="1673" alt="Screenshot 2021-08-25 at 17 50 54" src="https://user-images.githubusercontent.com/22457463/130832373-9e72dabb-4b7a-4a1c-a5a8-f1c96585d99b.png">

<img width="1449" alt="Screenshot 2021-08-25 at 17 51 04" src="https://user-images.githubusercontent.com/22457463/130832388-0d6719b9-77db-499e-be7d-035c0a7cc1d1.png">


This leads me to suspect the CORS issue is just a symptom of the problem, but not the actual problem.

**Comment by jasonhargrove** [comment]
thanks @RoniqueRicketts, I followed your trail and solved my CORS issue, which suddenly appeared after adding groups and cognito in the identity pool

- sign in, confirm access to REST api
- sign out
- create new group on cli, push
- in the cognito UI, add user to the new group
- sign in, observe CORS error on same api
- sign out
- remove user from group in cognito UI
- sign in, confirm access to api

solved this in the IAM ui

- go to Roles, filter for your group, click for detail
- attach policies: `amplify-xxxxxxxxx-xxxxxxxxx-xxxxxxxxx-authRole` which grants access to each route by method and resource

**Comment by jasonhargrove** [comment]
not sure but did not seem to matter which order I added auth or api as reported, so I dug a little further

```
tldr; after you add a group, you can use CLI to update your API, and attach the policy for that group
```

just now solved this with the cli, which is probably the intended flow

- after adding groups
- `amplify api update` choose `REST`
- select the api and path and follow the flow until you get to `restrict API access`
- select `restrict access by both` (auth and group)
- select `auth users only` and method
- select group(s)
- it will then ask you to confirm the method for each group
- amplify push

- confirm it added the policy to the group(s) in the flow via IAM ui
- ensure test user is in an impacted group
- sign out
- sign in, confirm access to resource

**Comment by pepso** [comment]
@josefaidt I strongly suggest this is not a documentation issue, but real technical problem with amplify-cli. Do I create a new issue, or can you reopen this and triage further as per my below details?

1. I have an API originally created with amplify-cli version ~4.20 with 10 or so paths all working fine and dandy
2. I was able to upgrade amplify-cli up to 4.52.0 version with manual workaround of parameters.json file clean up, and also getting the auth/unAuth roles recreated as per 4.50.2 minor version upgrade (which fixes api migration to new style of configuring the api gateway)
3. I created a new api path, a standard lambda as per my project (same "template" as other end points) and I got CORS all over. APIGateway does not invoke the lambda function associated with the newly created path
4. I decided to create a full "new" lambda example as per the amplify-cli walkthrough like this:
![image](https://user-images.githubusercontent.com/7467665/131324216-94579f31-bb59-45c5-86f2-0a6cb52b30e6.png)
5. Deployment all ok, but when I try to GET the end point (from remote server/localhost, or from deployed cloud stack [dev]) I get into CORS trouble.
6. OPTIONS returns 200, but GET fails:
![image](https://user-images.githubusercontent.com/7467665/131324370-1f5e8210-88a6-467a-95aa-48176104a051.png)

Please keep in mind every other API path in the same API Gateway end point continue to work, and be invoked as expected (created and deployed pre-4.50 amplify-cli). This impacts only the new paths created since 4.50.2 upgrade (which was supposed to fix the API migration issues).

I don't believe this to be documentation only issue, and it got to involve at least newly created out of box api + lambda walkthroughs update so the codegen works as expected.

Please advise what needs to be done to fix the disconnect between amplicy-cli created templates to working API path to Lambda invokes without CORS.

**Comment by pepso** [comment]
I fixed my issue and identified a breaking change in the way how amplify handles api calls/paths between versions:

Background:
In my web-app, every API call to my endpoints contain couple of variables in the body of the call (where I transfer some metrics about the user experience). In my backend I parse them with middleware using app.use(metricParsers()).

I found out that there is a change 
API paths created pre-4.50.2, GET calls with "body" go through API Gatway
API paths created post-4.50.2, GET calls fail at CORS validation if body is defined as part of the paylod

Basically I conclude that API Gateway configuration have become more strict with API updates in past couple of months.

**Comment by RoniqueRicketts** [comment]
@pepso as per the information sent by @jasonhargrove above you'll have to update the API to the groups you make. I surely believe the documentation should update on every update with beginner friendly content. That could minimize a lot of issue that I'm seeing on here. Right now there is also an issue with the API getting ```custom_headers``` but the documentation suggests how to send them but the information is incorrect as it doesn't match the actual reaction of the API on API Gateway.

**Comment by pepso** [comment]
@RoniqueRicketts thanks for the offer to help. I haven't created groups to my cognito, so it would be a surprise I would require an update to any groups.

My problem was solved literally by removing body:{} from the API call options. This is a call which works against GET paths created with earlier amplify-cli version: 
```
const APIName = 'apiname';
const URL = '/endpoint';
const options = { 
	headers: { "Content-Type": "application/json" }, 
	body: {},
	response: true 
};
const response = await API.get(APIName, URL, options);

```
And this is an API call which works post-4.50.2 amplify-cli created API paths:
```
const APIName = 'apiname';
const URL = '/endpoint';
const options = { 
	headers: { "Content-Type": "application/json" }, 
	response: true 
};
const response = await API.get(APIName, URL, options);

```

Using body as part of the GET call obviously is not according to the HTTP definition, but I just happend to use it to move some telemetry data from the app to the backend... Headers themselves have a limit (4-8Kb) depending on webserver of choice and its config) how much data you can transfer and for certain use cases the body is better place to transfer big telemetry payloads.

**Comment by josefaidt** [comment]
Hey @pepso 👋 are you sending the CORS headers from your Lambda? Are you seeing without adding those headers that the GET call succeeds with restriction enabled?

**Comment by pepso** [comment]
@josefaidt yes, I'm sending the CORS headers as per usual. 

If the body:{} is part of the payload, call never reaches lambda / lambda is not invoked by API gateway.

**Comment by josefaidt** [comment]
@pepso do you have your code in a public repository? I was not experiencing this with GET calls, shown with my example from https://github.com/aws-amplify/amplify-cli/issues/8019#issuecomment-905513096. Can you share what your Lambda code looks like?

**Comment by pepso** [comment]
@josefaidt I have remove the test end point which I used to narrow down the problem (from my private repo), but it was the box standard template generated by amplify-cli:

![image](https://user-images.githubusercontent.com/7467665/132022717-b6aa1885-c2e0-4302-98f5-ede11e486d83.png)

The example you mention in your comment differs significantly from the setup I have. In my setup `amplify api add` was done on amplify-cli version ~4.20 and the discussed `amplify api update` to add new path was with aplify-cli version 4.50.2

The lambda code does not get invoked as the payload is not going through the API Gateway.

If you want to replicate the problem as I have it, you probably need to start with older cli version by creating an API and migrateing it into newer versions step at a time. 

This is my upgrade path "story" if you want to replicate it:
I created my API originally on ~4.20 version. I was stuck on 4.29.2 for a long time due to change to init --force flag issue. When that was fixed on the CLI and I eventually got around to use the newly introduced flag, I upgraded to the 4.42 version. With this version I could not deploy any new APIs because CloudFormation updates were failing to auth/unAuth variable missing message. Later on I learned I need to manually remove auth/unAuth variables from parameters.json file as a workaround. Then, 4.50.2 actually fixed the auth/unAuth roles for my API Gateway. This is where I am at right now. For new api path created with current version, GET with body:{} fails to CORS with invalid signature without triggering the underlying lambda. GET with body:{} against the previously created api paths continue to work.

My current workaround is not to send body as part of GET payload to avoid this issue. I guess I stick with this until I have energy and time to try to upgrade 5.X onward.

@josefaidt a question about upgrade path to 5.x in mind... https://docs.amplify.aws/cli/migration/lambda-layers-update/

Can you confirm lambda layer one-way upgrade process is triggered also when deploying using amplify CD/init and I do NOT need to `amplify push` command from local dev box to cloud environments (dev/integration/production)?

**Comment by ranyakhemiri** [comment]
I had the same error and checked the Lambda logs, turns out the lambda was running out of memory. So I increased the memory and timeout and it solved the issue! 
Also enabling CORS through API Gateway console didn't solve the issue for some reason, but enabling CORS in the lambda code helped solve the issue : 
```
    return {
        statusCode: 200,
        headers: {
          "Access-Control-Allow-Origin": "*",
          "Access-Control-Allow-Headers": "*"
      }, 
        body: ... ,
    };
```
Hope this helps anybody struggling with the same issue!

**Comment by salmandjing** [comment]
> I had the same error and checked the Lambda logs, turns out the lambda was running out of memory. So I increased the memory and timeout and it solved the issue! Also enabling CORS through API Gateway console didn't solve the issue for some reason, but enabling CORS in the lambda code helped solve the issue :
> 
> ```
>     return {
>         statusCode: 200,
>         headers: {
>           "Access-Control-Allow-Origin": "*",
>           "Access-Control-Allow-Headers": "*"
>       }, 
>         body: ... ,
>     };
> ```
> 
> Hope this helps anybody struggling with the same issue!

Thank you so much. Took forever to find this solution.
